File: 03-classic-algorithms/searching/binarysearch/BinarySearch.java

/**
 * Copyright © https://github.com/jarry All rights reserved.
 * @author: jarryli@gmail.com
 * @version: 1.0
 */
import java.util.*;

// 标准非递归版本，传递左右区间指针
class BinarySearch {
  int search1(int[] arr, int target) {
    int left = 0;
    int right = arr.length - 1;
    int midIndex, mid;
    // left是左侧，right是最右侧。搜索区间长度小于1时停止
    while (left <= right) {
      midIndex = (left + right) / 2;
      mid = arr[midIndex];
      // 中间项等于目标项则返回下标
      if (target > mid) {
        // 大于中间项折半查找右侧
        left = midIndex + 1;
      } else {
        // 小于中间项折半查找左侧
        right = midIndex - 1;
      }
    }
    //  此时left是最左侧目标项
    if (left < arr.length && arr[left] == target) {
      return left;
    }

    return -1;
  }

  // 递归实现，移动左右区间指针
  int search2(int[] arr, int target, int left, int right) {
    right = right == -1 ? arr.length - 1 : right;
    int midIndex = (left + right) / 2;
    int mid = arr[midIndex];
    // 中间值等于查找项说明找到了，则返回中间项下标
    if (target == mid) {
      // 如果有重复项，返回第一个位置
      if (arr[left] == target) {
        return left;
      }
      return midIndex;
    }
    // 如果左侧与右侧相同，表面查找完毕，返回-1
    if (left >= right) {
      return -1;
    }
    if (target > mid) {
      // 折半右侧部分开始递归查找
      return this.search2(arr, target, midIndex + 1, right);
    } else {
      // 折半左侧部分开始递归查找
      return this.search2(arr, target, left, midIndex - 1);
    }
  }

  public static void main(final String args[]) {
    int[] arr = { 2, 3, 8, 10, 10, 13 };
    BinarySearch binarySearch = new BinarySearch();

    long startTime = System.currentTimeMillis();
    int result = binarySearch.search1(arr, 10); // 3
    System.out.println("binarySearch.search1(arr, 10) => result:" + result);
    result = binarySearch.search1(arr, 3); // 1
    System.out.println("binarySearch.search1(arr, 3) => result:" + result);
    result = binarySearch.search1(arr, 99); // -1
    System.out.println("binarySearch.search1(arr, 99) => result:" + result);
    System.out.println("\r\ntime:" + (System.currentTimeMillis() - startTime) + " ms.");

    startTime = System.currentTimeMillis();
    result = binarySearch.search2(arr, 10, 0, -1); // 3
    System.out.println("binarySearch.search2(arr, 10) => result:" + result);
    result = binarySearch.search2(arr, 3, 0, -1); // 1
    System.out.println("binarySearch.search2(arr, 3, 0, -1) => result:" + result);
    result = binarySearch.search2(arr, 99, 0, -1); // -1
    System.out.println("binarySearch.search2(arr, 99, 0, -1) => result:" + result);
    System.out.println("\r\ntime:" + (System.currentTimeMillis() - startTime) + " ms.");

  }
}

/**
jarry@jarrys-MacBook-Pro binarysearch % javac BinarySearch.java
jarry@jarrys-MacBook-Pro binarysearch % java BinarySearch      
binarySearch.search1(arr, 10) => result:3
binarySearch.search1(arr, 3) => result:1
binarySearch.search1(arr, 99) => result:-1

time:13 ms.
binarySearch.search2(arr, 10) => result:3
binarySearch.search2(arr, 3, 0, -1) => result:1
binarySearch.search2(arr, 99, 0, -1) => result:-1

time:1 ms.
 */

File: algorithmic-thinking/divide-and-conquer/binary-search/binary_search.js
/*
*
 * 二分查找 - 分治法的查询应用
 *
 * 前置条件：数组必须已排序
 *
 * 时间复杂度: O(log n)
 * 空间复杂度: O(1) 迭代版本，O(log n) 递归版本
*/

/*
*
 * 二分查找（迭代版本）
 * @param {number[]} arr - 已排序的数组
 * @param {number} target - 查找的目标值
 * @returns {number} 目标值的索引，如果不存在返回 -1
*/
function binarySearch(arr, target) {
    let left = 0, right = arr.length - 1;

    // 循环直到左右指针相遇, 找到目标值或确定不存在
    while (left <= right) {
        const mid = Math.floor(left + (right - left) / 2);
        // 检查中间元素是否是目标值
        if (arr[mid] === target) {
            return mid;
        // 如果中间元素小于目标值, 继续在右半部分搜索
        } else if (arr[mid] < target) {
            left = mid + 1;
        // 如果中间元素大于目标值, 继续在左半部分搜索
        } else {
            right = mid - 1;
        }
    }

    return -1;
}

/*
*
 * 二分查找（递归版本）
*/
function binarySearchRecursive(arr, left, right, target) {
    if (left > right) {
        return -1;
    }

    const mid = Math.floor(left + (right - left) / 2);
    // 检查中间元素是否是目标值
    if (arr[mid] === target) {
        return mid;
    } else if (arr[mid] < target) {
        // 如果中间元素小于目标值, 继续在右半部分搜索
        return binarySearchRecursive(arr, mid + 1, right, target);
    } else {
        // 如果中间元素大于目标值, 继续在左半部分搜索
        return binarySearchRecursive(arr, left, mid - 1, target);
    }
}

/*
*
 * 查找第一个 >= target 的元素位置（左边界）
*/
function binarySearchLeft(arr, target) {
    let left = 0, right = arr.length - 1;
    let result = -1;

    while (left <= right) {
        const mid = Math.floor(left + (right - left) / 2);

        if (arr[mid] >= target) {
            result = mid;
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }

    return result;
}

// 测试函数
console.log("=== 二分查找 ===\n");

const arr = [1, 3, 5, 7, 9, 11, 13, 15, 17, 19];
console.log("数组:", arr);
console.log();

// 测试用例1：迭代版本
console.log("1. 迭代版本:");
console.log("  查找 7:", binarySearch(arr, 7));
console.log("  查找 20:", binarySearch(arr, 20));
console.log();

// 测试用例2：递归版本
console.log("2. 递归版本:");
console.log("  查找 9:", binarySearchRecursive(arr, 0, arr.length - 1, 9));
console.log("  查找 2:", binarySearchRecursive(arr, 0, arr.length - 1, 2));
console.log();

// 测试用例3：查找边界
const arrDup = [1, 2, 2, 2, 3, 4, 5, 5, 5, 6];
console.log("3. 查找边界 (数组:", arrDup, "):");
console.log("  左边界 2:", binarySearchLeft(arrDup, 2));
console.log("  左边界 5:", binarySearchLeft(arrDup, 5));
console.log();

// 测试用例4：查找不存在的元素
console.log("4. 查找不存在的元素:");
console.log("  查找 100:", binarySearch(arr, 100));
console.log("  查找 0:", binarySearch(arr, 0));
